### PR TITLE
Fix crash in early data send with out-of-band PSK using AES CCM

### PR DESCRIPTION
**Pull request by raja-ashok**
TLSv1.3 client sending early data with out-of-band PSK using AES CCM ciphersuite crashes in `derive_secret_key_and_iv()` due to NULL pointer in `s->session->cipher`.

This needs backport to `1.1.1`.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by raja-ashok** [comment]
Added testcase to execute PSK out-of-band handshake for all 5 ciphersuites of TLSv1.3. And the testsuite `test_sslapi` gets executed 2 times. I feel 1st time is without FIPS and 2nd time is with FIPS.

2nd time execution of `test_sslapi` fails in the new testcase for Chacha20 Poly1305 ciphersuite. Can someone help me to find the issue here ?

```
        # Subtest: test_early_data_psk_with_all_5_ciphers
        1..5
        ok 1 - iteration 1
        ok 2 - iteration 2
        ok 3 - iteration 3
        ok 4 - iteration 4
        ok 5 - iteration 5
    ok 32 - test_early_data_psk_with_all_5_ciphers
....
        # Subtest: test_early_data_psk_with_all_5_ciphers
        1..5
        ok 1 - iteration 1
        ok 2 - iteration 2
        # ERROR: (bool) 'SSL_write_early_data(clientssl, MSG1, strlen(MSG1), &written) == true' failed @ test/sslapitest.c:3760
        # false
        # C0:CD:0C:C1:DB:7F:00:00:error:SSL routines::algorithm fetch failed:ssl/tls13_enc.c:617:
        not ok 3 - iteration 3
        ok 4 - iteration 4
        ok 5 - iteration 5
    not ok 32 - test_early_data_psk_with_all_5_ciphers
```

File: test/sslapitest.c [hunk 1]
```diff
@@ -3689,6 +3692,106 @@ static int test_early_data_psk(int idx)
...
+
+    if (!TEST_true(SSL_read_ex(serverssl, buf, sizeof(buf), &readbytes))
+            || !TEST_mem_eq(buf, readbytes, MSG2, strlen(MSG2)))
+        goto end;
+
+    testresult = 1;
+ end:
+    SSL_SESSION_free(sess);
+    SSL_SESSION_free(clientpsk);
+    SSL_SESSION_free(serverpsk);
```
Comment (kaduk) [inline test/sslapitest.c:3786]: (clientpsk and serverpsk are set by setupearly_data_test)

File: test/sslapitest.c [hunk 2]
```diff
@@ -3689,6 +3692,106 @@ static int test_early_data_psk(int idx)
...
+
+    /* Send some normal data from client to server */
+    if (!TEST_true(SSL_write_ex(clientssl, MSG2, strlen(MSG2), &written))
+            || !TEST_size_t_eq(written, strlen(MSG2)))
+        goto end;
+
+    if (!TEST_true(SSL_read_ex(serverssl, buf, sizeof(buf), &readbytes))
+            || !TEST_mem_eq(buf, readbytes, MSG2, strlen(MSG2)))
+        goto end;
+
```
Comment (kaduk) [inline test/sslapitest.c:3781]: Should we `SSL_shutdown()` clientssl and serverssl?

File: test/sslapitest.c [hunk 3]
```diff
@@ -7641,6 +7744,7 @@ int setup_tests(void)
     ADD_ALL_TESTS(test_early_data_skip_abort, 3);
     ADD_ALL_TESTS(test_early_data_not_sent, 3);
     ADD_ALL_TESTS(test_early_data_psk, 8);
+    ADD_ALL_TESTS(test_early_data_psk_with_all_5_ciphers, 5);
```
Comment (kaduk) [inline test/sslapitest.c:7747]: This name seems unlikely to age well.

File: test/sslapitest.c (hunk 1 above)
Comment (raja-ashok) [reply test/sslapitest.c:3786]: Yes. free is needed for all 3 session here, as reference is incremented for all these 3 by `setupearly_data_test()`

File: test/sslapitest.c (hunk 3 above)
Comment (raja-ashok) [reply test/sslapitest.c:7747]: Can I rename it as `test_early_data_psk_with_rfc8446_ciphersuites()` ?

File: test/sslapitest.c (hunk 2 above)
Comment (raja-ashok) [reply test/sslapitest.c:3781]: Yes, shutdown also should be done here.

File: test/sslapitest.c (hunk 3 above)
Comment (mattcaswell) [reply test/sslapitest.c:7747]: I'd just drop the `5` from the name, e.g. `test_early_data_psk_with_all_ciphers`.

**Comment by mattcaswell** [comment]
> 2nd time execution of test_sslapi fails in the new testcase for Chacha20 Poly1305 ciphersuite. Can someone help me to find the issue here ?

The FIPS module does not include ChaCha20-Poly1305 support - so that test should be skipped if the FIPS module is in use (i.e. if `is_fips` is set to 1).

**Comment by mattcaswell** [comment]
> TLSv1.3 client sending early data with out-of-band PSK using AES CCM ciphersuite crashes

Does this only impact AES CCM ciphersuite/s The fix seems to imply that this would be a problem for all out-of-band PSK ciphersuites.

**Comment by raja-ashok** [comment]
> Does this only impact AES CCM ciphersuite/s The fix seems to imply that this would be a problem for all out-of-band PSK ciphersuites.

Yes. It impacts only AES CCM ciphersuites. In `derive_secret_key_and_iv()` `cipher` is accessed to identify the tag length (16 or 8) for AES_CCM. For other ciphersuites `cipher` is not accessed.

File: test/sslapitest.c (hunk 1 above)
Comment (raja-ashok) [reply test/sslapitest.c:3789]: So no change required here.

File: test/sslapitest.c (hunk 2 above)
Comment (raja-ashok) [reply test/sslapitest.c:3784]: Added shutdown.

File: test/sslapitest.c (hunk 3 above)
Comment (raja-ashok) [reply test/sslapitest.c]: Changes done.

**Comment by mattcaswell** [comment]
Does this cherry-pick cleanly to 1.1.1 - or will we need a different PR there?

**Comment by raja-ashok** [comment]
> Does this cherry-pick cleanly to 1.1.1 - or will we need a different PR there?

I will raise a new PR for `1.1.1`. As cherry-pick is creating conflicts.

**Comment by mattcaswell** [comment]
Ok. I removed the 1.1.1 label from this PR for now.

File: test/sslapitest.c [hunk 4]
```diff
@@ -3689,6 +3692,111 @@ static int test_early_data_psk(int idx)
...
+        goto end;
+
+    testresult = 1;
+ end:
+    SSL_SESSION_free(sess);
+    SSL_SESSION_free(clientpsk);
+    SSL_SESSION_free(serverpsk);
+    clientpsk = serverpsk = NULL;
+    SSL_shutdown(clientssl);
+    SSL_shutdown(serverssl);
```
Comment (kaduk) [inline test/sslapitest.c:3792]: In my head the `SSL_shutdown()` calls went before the `end:` label, but it looks like we're not very consistent about where they go at the moment.  That is, the current location should be fine as-is.

Comment (raja-ashok) [reply test/sslapitest.c:3792]: Ohk. Looks like a NULL check is needed for this `SSL_shutdown` call at this place. I will add it.

Comment (raja-ashok) [reply test/sslapitest.c]: Fixed

**Comment by kaduk** [comment]
Thanks for adding the NULL checks; I reconfirm my +1.
@mattcaswell or @t8m should reconfirm that it still looks good from their end, and I will toggle approval:done off and on to re-set the timer.

**Comment by kaduk** [comment]
I'll go ahead and start merging

**Comment by kaduk** [comment]
> I'll go ahead and start merging

... except the repo is frozen for alpha2.  Too distracted to remember, I guess.

**Comment by kaduk** [comment]
Merged to master, so closing.
Thanks!

**Pull request closed without merging**